### ISSUE TITLE
Misleading assertion message

### ISSUE THREAD

**Issue by rashchedrin** [Z3Prover/z3#3439]
Here
https://github.com/Z3Prover/z3/blob/37bc4a4407d1cd4eab9de1e26c546aba5d4e8d26/src/api/python/z3/z3.py#L1408
Assertion checks that val is expression, but message says that "True" and "False" are correct values as well, which is not true. `z3.is_expr(True)` is False. As well as `z3.is_expr(False)` is False. 

I don't know why my code hit this assertion (sometimes it does, sometimes it doesn't), but replacing `True` to `z3.BoolVal(True)` in my code solved my problem (maybe that's because I use `import z3` instead of `from z3 import *`). And before I replaced `True` to `z3.BoolVal(True)`, I was receiving an exception with message 
`True, False or Z3 Boolean expression expected. Received True`
Which is not informative.

**Comment by NikolajBjorner** [comment]
The exception message is intended to cover also the code right above line 1408.
There it checks for whether the argument is a bool. I can't say why that code doesn't apply to your situation. It works well here and there is even a doc test to check this case.

**Comment by rashchedrin** [comment]
I've figured it out. My problem arise from the fact that I used values from numpy bool array, and numpy bool arrays return objects of type `numpy.bool_`, which is not `bool`. So `isinstance(np.array([True])[0], bool) == False`. Here is the minimal example:
```
import numpy as np
import z3
s = z3.Solver()
s.add(z3.Bool('v0') == np.array([True])[0])
```
it rises 
```
Z3Exception: True, False or Z3 Boolean expression expected. Received True
```
I tried changing `if isinstance(val, bool):` to `if val == True or val == False:` on line https://github.com/Z3Prover/z3/blob/37bc4a4407d1cd4eab9de1e26c546aba5d4e8d26/src/api/python/z3/z3.py#L1404 , but then tests fail.
Still, it is possible to make the exception message more meaningful by printing type of value. For example message 
```
Z3Exception: True, False or Z3 Boolean expression expected. Received True of type <class 'numpy.bool_'>
```
will be more useful.
